Split GitHub Actions and TuxSuite by toolchain version

nathanchance opened the pull request:
This series changes how our GitHub Actions and TuxSuite files are structured to make it easier to triage build failures and see the overall health of the trees we test and to avoid hitting GitHub Actions' job limit in our matrices.

The first two patches move the reading of generator.yml and writing of the GitHub Actions workflow and TuxSuite files into the Python scripts, rather than using stdin and stdout so that we can easily write to multiple files.

The third patch does the actual split by checking what LLVM versions the tree is being tested with and generating a file based on just those builds. `generate.sh` and `check-patches.sh` are adjusted to account for this.

Finally, the fourth and fifth patch does the regeneration and updates the README.

nathanchance commented:
Need to run `yapf` and fix the `--patch-series` handling.

nickdesaulniers commented on generate_tuxsuite.py:
```diff
@@ -30,14 +30,18 @@ def get_repo_ref(config, tree_name):
 def emit_tuxsuite_yml(config, tree):
+    tuxsuite_yml = "tuxsuite/{}.tux.yml".format(tree)
+    orig_stdout = sys.stdout
+    f = open(tuxsuite_yml, "w")
```
aren't `with` blocks more canonical python?

nickdesaulniers replied:
i.e.
```python
with open(tuxsuite_yml, "w") as f:
    <indent rest of code that needs f in scope>
```
then you can drop the explicit close method call.

nathanchance replied:
Sure, if that is what you prefer. I did it this way as it is less indenting.

nickdesaulniers commented on generate_workflow.py:
```diff
@@ -196,10 +196,15 @@ def print_builds(config, tree_name):
         workflow["jobs"].update(check_logs_allconfigs)
 
+    orig_stdout = sys.stdout
+    f = open(github_yml, "w")
```
can we use a `with` block here, too?

nickdesaulniers commented on generate_tuxsuite.py:
```diff
@@ -59,11 +54,11 @@ def emit_tuxsuite_yml(config, tree):
     allconfigs = []
     for build in config["builds"]:
-        if build["git_repo"] == repo and build["git_ref"] == ref:
+        if build["git_repo"] == repo and build["git_ref"] == ref and build[
```
the wrapping here is a bit jarring. Can the line end with `and` so that `build["llvm_version"]` starts the next line?

nathanchance replied:
This is just `yapf` does.

I could do this if you would prefer?

```diff
diff --git a/generate_tuxsuite.py b/generate_tuxsuite.py
index 7b5ecab..da1b43c 100755
--- a/generate_tuxsuite.py
+++ b/generate_tuxsuite.py
@@ -54,8 +54,9 @@ def emit_tuxsuite_yml(config, tree, llvm_version):
     distribution_configs = []
     allconfigs = []
     for build in config["builds"]:
-        if build["git_repo"] == repo and build["git_ref"] == ref and build[
-                "llvm_version"] == llvm_version:
+        if build["git_repo"] == repo and \
+           build["git_ref"] == ref and \
+           build["llvm_version"] == llvm_version:
             arch = build["ARCH"] if "ARCH" in build else "x86_64"
             if llvm_version == max_version:
                 toolchain = "clang-nightly"
```

nickdesaulniers replied:
that looks nicer; won't `yapf` undo that though without annotations?

nathanchance replied:
Does not appear so. I've included it for the next push.

nickdesaulniers commented on generate_workflow.py:
```diff
@@ -163,16 +157,18 @@ def get_steps(build, build_set):
     check_logs_allconfigs = {}
     for build in config["builds"]:
-        if build["git_repo"] == repo and build["git_ref"] == ref:
+        if build["git_repo"] == repo and build["git_ref"] == ref and build[
```
line wrapping :(

nickdesaulniers commented on utils.py:
```diff
@@ -136,6 +136,21 @@ def get_build():
+
+
+def get_llvm_versions(config, tree_name):
+    llvm_set = set()
```
s/llvm_set/llvm_versions/ or just `versions` (we know it's a set, avoid hungarian notation)

nathanchance replied:
Sounds good, I'll do `llvm_versions`.

nickdesaulniers commented on generate.sh:
```diff
@@ -19,8 +19,8 @@ done
-    ./generate_tuxsuite.py <generator.yml "${BRANCH}" >tuxsuite/"${BRANCH}".tux.yml
-    ./generate_workflow.py <generator.yml "${BRANCH}" >.github/workflows/"${BRANCH}".yml
+    ./generate_tuxsuite.py "${BRANCH}" >tuxsuite/"${BRANCH}".tux.yml
+    ./generate_workflow.py "${BRANCH}" >.github/workflows/"${BRANCH}".yml
```
or can be in a child commit. doesn't matter that much to me.

nathanchance replied:
Thanks, good point. I will update the README as part of the two commits that rework the reading and writing.

nathanchance commented on generate_tuxsuite.py:
```diff
@@ -30,14 +30,18 @@ def get_repo_ref(config, tree_name):
+    tuxsuite_yml = "tuxsuite/{}.tux.yml".format(tree)
+    orig_stdout = sys.stdout
+    f = open(tuxsuite_yml, "w")
+    sys.stdout = f
```
Yes, I figured it was easier to just keep the prints but redirect them to the file.

nathanchance commented:
Thank you for the review! I have gone ahead and addressed all your comments via `git commit --fixup`, which I'll rebase and autosquash once you confirm you are happy with the changes.

nickdesaulniers reviewed (approved):
lgtm nice work! 🌊 👋
